### ISSUE TITLE
Help understanding your work 

### ISSUE THREAD

**Issue by ShowStopperTheSecond** [interestingzhuo/PNPloss#1]
Hi there :) 
I'm trying to replicate your work to fully understand it. But I have difficulties following your what you did. I have the following function following what you described in your paper:

```


def sigmoid(tensor, temp=0.01):
    exponent = -tensor / temp
    y = 1.0 / (1.0 + torch.exp(exponent))
    return y

def loss(labels, x, b=2, alpha=1, anneal=1e-4, variant="PNP-O"):
    # Cosine Similarity 
    x = x.squeeze()
    x = x / torch.norm(x, dim=1, keepdim=True)
    sim = x @ x.t()

    # Creating the lable-wise mask (each row specifies the instances that are from the same class)
    mask = labels[:, None]==labels[None, :]


    # Creating the lable-wise mask (each row specifies the instances that are from the same class)
    mask = labels[:, None]==labels[None, :]
    
    # Finding the positive-negative samples (rows specify negative samples and column positive ones)
    # Each row is the location of negative positive samples 
    np_mask = ~mask[:, :, None] & mask[:, None, :]
    
    # Using Sigmoid function to find the instances where negative samples are greater than the positive ones
    sim_dist = sim[:, :, None] - sim[:, None, :]
    sig_dist = sigmoid(sim_dist)
    
    
    # Selecting only the negatives before positives 
    negatives_before_positives = sig_dist * np_mask
    
    # Finding the number of positive instances  for normalization
    n_positives = mask.sum(-1)
    
    # Computing the R(i, S_N)
    R_I_SN = negatives_before_positives.sum(-1)
    
    # Normalizing by the number of positive samples
    PNP_O = R_I_SN/n_positives
    
    loss = PNP_O[torch.isfinite(PNP_O)].sum()
                
    return loss


```
But what happens is that the loss will quickly fall to zero and the training fails. Am I really following exactly what you did in your paper? What am I failing to understand. I tried to train an test network on CIFAR100 to create good embeddings. But after a couple of epochs the loss is zero. Here is the Kaggle notebook ( https://www.kaggle.com/javidtheimmortal/pnplosspytorch). I'd appreciate if you could help me understand where am I falling short. 
Thanks :)

**Comment by interestingzhuo** [comment]
you can try the implement in https://github.com/KevinMusgrave/pytorch-metric-learning

**Comment by ShowStopperTheSecond** [comment]
Thanks it's just what I need :) I'll see how it's done there
